Implemented changed methods

NullConvergence opened the pull request:
I implemented the changed methods routines, checking if methods were removed, added or have any difference in ``nloc``.

More tests will follow after we solve the following issue:

``nloc`` is not a good metric to check if 2 methods are the same. 
Which would be a good way to check this? The number of tokens? A combination of multiple measurements from lizard? 😄

ishepard commented:
Hi @NullConvergence!
Thanks for the PR!

I was thinking to use a more reliable approach to detect changed methods. 

From the diff, we now which lines have been modified (literally we can use the line number).
From lizard, we know for each method the starting and ending line. 

Hence, we can just check, for each modified line (deleted by looking at the old source code, added by looking at the new code) whether it belongs to a method.

Does it make sense? I think in this way it's more reliable than just looking at the names.

NullConvergence commented:
I think it makes sense, although it will be slower. I pushed a small update + 1 test. 
Let me know if this is what you thought of and I can add more tests.

ishepard commented on pydriller/domain/commit.py:
```diff
@@ -254,18 +255,62 @@ def methods(self) -> List[Method]:
+            y for x in deleted for y in old_methods if x[0] <= y.end_line and x[0] >= y.start_line
+        ])
+
+        return list(methods_changed_new.union(methods_changed_old))
```
The union will probably fail for changed methods (but not for added or deleted). 
The reason is that for the same method, the object `Method` in `methods_changed_new` is different from the object `Method` in `methods_changed_old`. Hence we will have the same method twice.
This is just a supposition, we need more testing

ishepard replied:
If this is the case, just consider one of the 2

NullConvergence replied:
Yes, I was expecting this, but we have to decide which would be the Method's attribute to filter for. Is name suitable for detecting duplicates in this case?

NullConvergence replied:
This is solved by adding  __hash__ and __eq__ to the 'Method' class. We should check the unique attributes there.

NullConvergence commented on pydriller/domain/commit.py:
```diff
@@ -254,18 +255,62 @@ def methods(self) -> List[Method]:
+        deleted = self.diff_parsed['deleted']
+
+        methods_changed_new = set([
+            y for x in added for y in new_methods if x[0] <= y.end_line and x[0] >= y.start_line])
```
Nice! I didn't knew this works out of the box.

NullConvergence commented:
I implemented an __eq__ and __hash__ methods for the Method class, which allow us to compare two methods (so union works perfectly). 

Currently I use the name + parameters (parameters help with name overloading). I also added various tests, but not the actual repo zip. How should I do that? Just zip the test_repos and upload them?

ishepard commented:
If you changed something in the `test-repos` folder yes, otherwise no. 

From what I saw, you didn't change anything in that folder. 

Anyway, some tests are failing in the CI, could you please check why and fix them?

Thanks!

PS: no worries for the green light from the CI, even if the tests pass the CI is gonna fail because of codecov. They should fix this error soon, but in the meanwhile don't look at the result of CI, just check if the tests pass 😄

NullConvergence commented:
The test fail because I added several commits in the A.java file. I will zip the folder and upload it to git.

NullConvergence commented:
The tests pass after the upload :shipit:!

ishepard commented:
Beautiful! Thanks!

I will message you when the new version will be shipped with this feature!

NullConvergence commented:
Thanks! Cheers.
